### PR TITLE
Refactor user caching to improve readability and fix threadsafety issues

### LINKED ISSUE #10852: Crash after entering the "currently playing" tab
**Screenshots or videos showing encountered issue:** 
https://youtu.be/gL_3urGVWlI

**osu!lazer version:** 
2020.1114.0

**Logs:**
[database.log](https://github.com/ppy/osu/files/5542552/database.log)
[network.log](https://github.com/ppy/osu/files/5542553/network.log)
[performance.log](https://github.com/ppy/osu/files/5542554/performance.log)
[runtime.log](https://github.com/ppy/osu/files/5542555/runtime.log)
[updater.log](https://github.com/ppy/osu/files/5542556/updater.log)

**Computer Specifications:** 
OS: windows 10
CPU: i5-4440
iGPU: intel hd graphics 4600
RAM: 8 GB

### PR DESCRIPTION
**Pull request by smoogipoo**
Closes https://github.com/ppy/osu/issues/10852
Supersedes and closes https://github.com/ppy/osu/pull/10861

With the superseded PR, I found it very hard to wrap my head around:
1. Request queues and nested user queues.
2. Task continuations.
3. The nested locking.

So this PR attempts to simplify this logic - it stores one queue of `(userId, TaskCompletionSource)` which can grow indefinitely even if there's duplicates, and the `TaskCompletionSource` for each is what's awaited on by the cache. Requests grab as many as 50 users from this queue that have not been fulfilled/are not duplicates.

### CODE REVIEW

**Comment by peppy** [comment]
Seems pretty good readability wise.

Why not make `pendingUserTasks` a queue? It would save the manual `RemoveAt` etc.

Also the way you have it you are running the newest queries first, rather than the oldest (also would be fixed by using standard queue operations).

**Review by peppy — CHANGES_REQUESTED** [review]
as proposed

File: osu.Game/Database/UserLookupCache.cs [hunk 1]
```diff
@@ -15,103 +14,73 @@ namespace osu.Game.Database
...
-                    // append the new request on to the last to be executed.
-                    pendingRequest = pendingRequest.ContinueWith(_ => work());
-                    pendingRequestConsumedIDs = false;
+                    (int id, TaskCompletionSource<User> task) next = pendingUserTasks.Dequeue();
+
+                    // Perform a secondary check for existence, in case the user was queried in a previous batch.
+                    if (CheckExists(next.id, out var existing))
+                        next.task.SetResult(existing);
+                    else
+                        userTasks.Add(next);
```
Comment (peppy) [inline osu.Game/Database/UserLookupCache.cs:61]: Not 100% on the necessity of the `CheckExisting` call you have, but there's also of requesting the same user in the lookup twice (which was handled on my PR).

To fix this, we could either:

- Store the IDs in a separate `HashSet` alongside `userTasks` (ensures always sending a request for 50 users).
- call Distinct in the `userTasks.Select(t => t.id)` below (simpler to implement, but requests may be short of 50. probably doesn't matter)

leaning towards the second option.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Database/UserLookupCache.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Game/Database/UserLookupCache.cs:61]: At the point where we're calling `Distinct`, we may as well just store a temporary hashset along with the main list and keep the up-to-50 users. Fine with that?

Comment (peppy) [reply osu.Game/Database/UserLookupCache.cs:61]: ok

**Comment by bdach** [comment]
I like this version a lot as written (forgot `TaskCompletionSource` was a thing, which is a big shame, as it's really useful...) About the only thing I'm not 100% about is this potentially-quadratic double loop:

https://github.com/ppy/osu/blob/85b0f714670215fc85eec8215f393674ccfd062a/osu.Game/Database/UserLookupCache.cs#L85-L87

although given that the user count is capped to 50 maybe not really a huge deal (2.5k accesses to elements in the worst case, barring pathological cases with multiple duplicates).

**Comment by peppy** [comment]
yeah, considered suggesting a dictionary to avoid that, but it comes at added complexity for a relatively uncommon execution path.

**Comment by peppy** [comment]
Have implemented using dictionary, @smoogipoo cherry pick if you agree with it i guess.

https://github.com/peppy/osu/commit/009d666241fb7adfd1c1f11d7603478dbd9166b0

**Review by peppy — APPROVED** [review]

**Pull request merged**